daffodil-runtime2/src/main/scala/org/apache/daffodil/runtime2/generators/CodeGeneratorState.scala

```diff
@@ -105,41 +129,178 @@ class CodeGeneratorState {
+
+    val dispatchField = choiceDispatchField(context)
+    if (dispatchField.nonEmpty) {
+      val C = localName(context)
+      val declaration =
+        s"""    size_t      _choice; // choice of which union field to use
+           |    union
+           |    {""".stripMargin
+      val erdDef =
+        s"""static const ERD _choice_$erd = {
+           |    {
+           |        NULL, // namedQName.prefix
+           |        "_choice", // namedQName.local
+           |        NULL, // namedQName.ns
+           |    },
+           |    CHOICE, // typeCode
+           |    0, NULL, NULL, NULL, NULL, NULL, NULL
+           |};
+           |""".stripMargin
+      val offsetComputation = s"    (const char *)&${C}_compute_offsets._choice - (const char *)&${C}_compute_offsets"
+      val erdComputation = s"    &_choice_$erd"
+      val initStatement = s"    instance->_choice = NO_CHOICE;"
+      val initChoiceStatement =
+        s"""    int64_t key = rootElement->$dispatchField;
+           |    switch (key)
+           |    {""".stripMargin
+      val parseStatement =
+        s"""    instance->_base.erd->initChoice(&instance->_base, rootElement());
+           |    switch (instance->_choice)
+           |    {""".stripMargin
+      val unparseStatement =
+        s"""    instance->_base.erd->initChoice(&instance->_base, rootElement());
+           |    switch (instance->_choice)
+           |    {""".stripMargin
+
+      erds += erdDef
+      structs.top.declarations += declaration
+      structs.top.offsetComputations += offsetComputation
+      structs.top.erdComputations += erdComputation
+      structs.top.initStatements += initStatement
+      structs.top.initChoiceStatements += initChoiceStatement
+      structs.top.parserStatements += parseStatement
+      structs.top.unparserStatements += unparseStatement
+    }
+  }
+
+  def addAfterSwitchStatements(): Unit = {
+    if (structs.top.initChoiceStatements.nonEmpty) {
+      val declaration = s"    };"
+      val initChoiceStatement =
+        s"""    default:
+           |        instance->_choice = NO_CHOICE;
+           |        break;
+           |    }
+           |
+           |    if (instance->_choice != NO_CHOICE)
+           |    {
+           |        const size_t choice = instance->_choice + 1; // skip the _choice field
+           |        const size_t offset = instance->_base.erd->offsets[choice];
+           |        const ERD *  childERD = instance->_base.erd->childrenERDs[choice];
+           |        InfosetBase *childNode = (InfosetBase *)((const char *)instance + offset);
+           |        childNode->erd = childERD;
+           |        return true;
+           |    }
+           |    else
+           |    {
+           |        return false;
+           |    }""".stripMargin
+      val parseStatement =
+        s"""    default:
+           |        pstate->error_msg = "node's _choice field was not initialized during parsing";
```

Comment:
I already made this comment in the example code. But this should be a different kind of error. We need to distinghish errors that come from users making mistakes writing schemas, and failures of the implementation to correctly compile or run the schema.
The latter are fatal errors and the right thing is probably to fail and stop processing. (Fail in simplest way possible.)

Reply:
This kind of error - choice not getting initialized properly. Is not a DFDL "processing error" in the formal DFDL sense that it should cause parser backtracking or result in a "Parse Error" message.
It's an internal code invariant failure. It indicates the runtime2 system has a bug and is not initializing things properly. 
These kinds of errors are strongly distinguished throughout the Daffodil code base. We should continue that in the runtime2 code. 
I think you should just call a global invariantFailed("bug - choice field not initialized") function. That should probably just abort/throw/longjmp out and end processing. In principle, once the runtime2 code is correct, this code path should not be even be reachable.

Reply:
I did the following tests:

```shell
$ cd daffodil-runtime2/src/test/resources/org/apache/daffodil/runtime2
$ daffodil generate c -s nested.dfdl.xsd -r NestedUnion
$ cp nested.dfdl.xsd c/
$ cp nested_union_parse_4.dat c/parse.dat
$ cp nested_union_unparse_4.xml c/unparse.xml
$ cd c
$ make tests
./daffodil parse parse.dat -o test_unparse.xml
xmldiff unparse.xml test_unparse.xml

./daffodil unparse unparse.xml -o test_parse.dat
diff parse.dat test_parse.dat
$ : Also check Daffodil runtime1...
$ daffodil parse -s nested.dfdl.xsd -r NestedUnion -o test_unparse.xml parse.dat # No errors
$ xmldiff unparse.xml test_unparse.xml # No errors
$ daffodil unparse -s nested.dfdl.xsd -r NestedUnion -o test_parse.dat unparse.xml # No errors
$ diff parse.dat test_parse.dat # No errors
$ : Now change the choice dispatch key in parse.dat...
$ emacs parse.dat # change 4th byte from ^D to ^@
$ make parse-test
./daffodil parse parse.dat -o test_unparse.xml
./daffodil: Parse error: no match between choice dispatch key and any branch key
make: *** [Makefile:42: parse-test] Error 1
$ daffodil parse -s nested.dfdl.xsd -r NestedUnion -o test_unparse.xml parse.dat
[error] Parse Error: Choice dispatch key (0) failed to match any of the branch keys.
Schema context: choice[1] Location line 71 column 12 in file:/home/interran/apache/incubator-daffodil-runtime2/daffodil-runtime2/src/test/resources/org/apache/daffodil/runtime2/c/nested.dfdl.xsd
Data location was preceding byte 4
$ : Now change the choice dispatch key in unparse.xml as well....
$ emacs unparse.xml # change <tag>4</tag> to <tag>0</tag>
$ make unparse-test
./daffodil unparse unparse.xml -o test_parse.dat
./daffodil: Walk error: no match between choice dispatch key and any branch key
make: *** [Makefile:46: unparse-test] Error 1
$ daffodil unparse -s nested.dfdl.xsd -r NestedUnion -o test_parse.dat unparse.xml # No errors
$ diff parse.dat test_parse.dat # No errors, 4th byte is zero in both files
```

My point is that this line in CodeGeneratorState which we are talking about makes sure c-daffodil behaves the same as scala-daffodil.  Both report a processing error saying there is no match between the choice dispatch key and any branch key when parsing the parse.dat file.  You can say that c-daffodil shouldn't be more strict than scala-daffodil when unparsing (it reports a no match processing error on the modifed unparse.xml while scala-daffodil silently writes out the same modified parse.dat) but that's a different part of code than this one.